### DISCUSSION TITLE
Possible to use alias/function for the shells used by fzf?

### DISCUSSION THREAD

**Discussion by enoryw** [junegunn/fzf#3993] [Q&A] [answered]
I have the following bash shell script (I would like to convert to a zsh shell function):

    #!/usr/bin/env bash

    cmd() {
      fd -t d --hidden --no-ignore --follow \
        --ignore-file "$HOME/.config/fd/ignore" \
        --ignore-file "$HOME/.config/fd/cd-fzf.fdignore" \
        . "$@" 
    }

    export -f cmd

    selection=$(cmd "$HOME" /etc /media \
      /data /data2 | SHELL=/usr/bin/bash fzf --preview 'tree -C {}' \
      --header 'cd to selection (Ctrl-r to toggle filtering for non-media dirs only)' \
      --prompt 'Dirs > ' \
      --bind 'ctrl-r:transform:[[ ! $FZF_PROMPT == "Dirs > " ]] &&
      echo "change-prompt(Dirs > )+reload(cmd "$HOME" /etc /media /data /data2)" ||
      echo "change-prompt(Dirs (media excluded) > )+reload(cmd /etc"$HOME/dev")"')

    if [ -n "$selection" ]; then
      cd "$selection" && n "$selection"
    fi

Here it relies on `export -f cmd` simplify the code so I don't have the repeat the same `fd` command and arguments. Is it possible to have an equivalent without exporting the function? AFAIK exporting a function in shells like bash/zsh is considered broken and not supported but I'm not sure of any alternatives. Can the above be restructured to get the same effect or is hardcoding and repeating the commands/arguments the only "proper" way?

**Comment by LangLangBart** [comment, marked as answer]
> Can the above be restructured to get the same effect

yes. 

[shell script - what is the zsh equivalent of bash's export -f - Unix &amp; Linux Stack Exchange](https://unix.stackexchange.com/questions/59360/what-is-the-zsh-equivalent-of-bashs-export-f)

---

Most common is to store the function in a variable name.

```sh
cmd='fd -t d --hidden --no-ignore --follow --ignore-file "$HOME/.config/fd/ignore" --ignore-file "$HOME/.config/fd/cd-fzf.fdignore" .'

selection=$(fzf --preview 'tree -C {}' \
  --prompt 'Dirs > ' \
  --bind "start:reload:$cmd '$HOME' /etc /media /data /data2" \
  --bind "ctrl-r:transform:[[ ! \$FZF_PROMPT == 'Dirs > ' ]] &&
    echo \"change-prompt(Dirs > )+reload($cmd '$HOME' /etc /media /data /data2)\" ||
    echo \"change-prompt(Dirs (media excluded) > )+reload($cmd /etc '$HOME/dev')\"")

if [ -n "$selection" ]; then
  cd "$selection" && n "$selection"
fi
```

**PRO**
- The function and its use case are close together, and the implementation should work in both bash and zsh.

**CONTRA**
- Functions crammed into a string might cause issues with proper quoting and make them harder to read due to missing syntax highlighting and not triggering shellcheck.

---


Another way is using `.zshenv`[^1], as it is the only initialization file that is loaded with every `zsh` session, except when used with the `-f/--no-rcs` flag. 

```sh
# Define a function in your  "${ZDOTDIR:-$HOME}/.zshenv"
random_word() {
  perl -e 'srand; rand($.) < 1 && ($line = $_) while <>; print $line;' "${1:-/usr/share/dict/words}"
}
```

```sh
# this works, it will find the function
fzf --bind "start:reload:random_word" --preview 'typeset -f'

# this won't work, the function will not be found; the '--with-shell' flag came with '0.51.0' (see CHANGELOG.md)
fzf --bind "start:reload:random_word" --preview 'typeset -f' --with-shell 'zsh --no-rcs -c'
```

**PRO**
- Functions can be written normally and benefit from shell formatting, syntax highlighting, and shellcheck in the editor to identify potential issues and improve readability.

**CONTRA**
- The function and its use case are likely spread across different files.


[^1]: [Z-Shell Frequently-Asked Questions](https://zsh.sourceforge.io/FAQ/zshfaq03.html#l20)

**Reply by rieje** [reply]
I'm trying to adapt OP's code with the solution to use function as variable name and have the following (this is a zsh autoload function):

    local fj_dir=$(echo $HOME/.firejail/*/Downloads/)

    local cmd="fd -t d --hidden --follow ."

    local user_dir="$HOME /tmp /media"
    local sys_dir="/usr /etc /var/cache/pacman"

    # Make available to fzf child shell processes
    export cmd fj_dir user_dir sys_dir

    selection=$({ printf %s\\n ${=fj_dir} & ${=cmd} ${=user_dir} } |
      SHELL=/usr/bin/bash fzf --scheme=path --preview 'tree -C {}' \
      --header 'cd to selection (Ctrl-r to toggle for sys dirs)' \
      --prompt 'Dirs (user) > ' \
      --bind 'ctrl-r:transform:[[ ! $FZF_PROMPT == "Dirs (user) > " ]] &&
      echo "change-prompt(Dirs (user) > )+reload({ printf %s\n $fj_dir & $cmd $user_dir ;})" ||
      echo "change-prompt(Dirs (system) > )+reload($cmd $sys_dir)"')

On initial fzf, it prints the correct list of items and on reload to toggle the other set of directories, it's correct. But when subsequently toggling back to the initial set of directories results in `$fj_dir` being printed as one one line:

    `/home/riejew/.firejail/firefox-a/Downloads/n/home/riejew/.firejail/firefox-b/Downloads/n/home/riejew/.firejail/firefox-c/Downloads/n/home`...

instead of being split as separate lines for fzf. Any ideas? The culprit is `reload({ printf %s\\n $fj_dir & $cmd $user_dir ;})` (I assume something is wrong with the syntax to make the items delimit by "n" instead of newline. The reason I have `{ printf %s\\n $fj_dir & ...` is because I want to include the search directories `$fj_dir` as part of the result but [fd does not support that](https://github.com/sharkdp/fd/issues/364#issuecomment-2116942923) (`find` does, but I need to expand this function in the future and want `fd` to work here).

Any ideas? Also with the dir variables as strings I'm not sure how to support potential dirs with spaces.

Thanks.
